File: examples/typescript/operations/dag/run-if.ts
import {
  DurableContext,
  TaskStatus,
  withDurableExecution,
} from "@aws/durable-execution-sdk-js";

type RefundOutcome = {
  refund: TaskStatus | undefined;
  notify: TaskStatus | undefined;
};

export const handler = withDurableExecution(
  async (_event: unknown, context: DurableContext): Promise<RefundOutcome> => {
    const result = await context.dag("conditional-refund", (dag) => {
      const order = dag.step("load-order", [], async () => ({ total: 0 }));

      const refund = dag.step(
        "issue-refund",
        [order],
        async (deps) => `refunded ${deps["load-order"].total}`,
        { runIf: (deps) => deps["load-order"].total > 0 },
      );

      dag.step("notify", [refund], async (deps) => `sent: ${deps["issue-refund"]}`);
    });

    return {
      refund: result.getStatus("issue-refund"),
      notify: result.getStatus("notify"),
    };
  },
);

File: examples/python/operations/dag/run-if.py
from aws_durable_execution_sdk_python import (
    DagContext,
    DagResult,
    DurableContext,
    durable_execution,
)


@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    def register(dag: DagContext) -> None:
        order = dag.step(lambda deps, sc: {"total": 0}, name="load-order")
        refund = dag.step(
            lambda deps, sc: f"refunded {deps['load-order']['total']}",
            deps=[order],
            name="issue-refund",
            run_if=lambda deps: deps["load-order"]["total"] > 0,
        )
        dag.step(
            lambda deps, sc: f"sent: {deps['issue-refund']}",
            deps=[refund],
            name="notify",
        )

    result: DagResult = context.dag(register, name="conditional-refund")
    return {
        "refund": result.get_status("issue-refund").value,
        "notify": result.get_status("notify").value,
    }
